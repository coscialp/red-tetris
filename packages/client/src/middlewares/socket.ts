import { Socket } from "socket.io-client";

export const socketMiddleware = (socket: Socket) => (params) => (next) => (action) => {
  const { dispatch, getState } = params
  const { type } = action

  switch (type) {
    case "socket/connect":
  }
};