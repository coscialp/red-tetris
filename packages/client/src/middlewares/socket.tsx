import { io } from "socket.io-client";
import { BACK_URL } from "../utils/ip.ts";

export enum SocketActionTypes {
  CONNECT = "socket/connect",
  DISCONNECT = "socket/disconnect",
  EMIT = "socket/emit",
  ON = "socket/on",
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const socketMiddleware = (store: any) => (next: any) => (action: any) => {
  const { type } = action;
  const { socket, setSocket } = store.getState().rootReducer;

  switch (type) {
    case SocketActionTypes.CONNECT:
      if (!socket) {
        setSocket(io(`ws://${BACK_URL}:3001`, { transports: ['websocket'] }));
      }
      break;
    case SocketActionTypes.DISCONNECT:
      if (socket) {
        socket.disconnect();
      }
      break;
    case SocketActionTypes.EMIT:
      if (socket) {
        socket.emit(action.event, action.data);
      }
      break;
    case SocketActionTypes.ON:
      if (socket) {
        socket.on(action.event, action.callback);
      }
      break;
    default:
      break;
  }

  return next(action);
};