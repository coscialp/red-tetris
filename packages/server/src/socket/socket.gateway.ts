import {
  ConnectedSocket,
  MessageBody,
  OnGatewayConnection,
  OnGatewayDisconnect,
  SubscribeMessage,
  WebSocketGateway,
  WebSocketServer,
} from "@nestjs/websockets";
import { Socket } from "socket.io";
import Player from "../logics/Player";
import Game from "../logics/Game";
import { PieceFactory } from "../logics/factory";
import { Piece } from "src/logics/Piece";

@WebSocketGateway(3001, { transports: ["websocket"] })
export class SocketGateway implements OnGatewayConnection, OnGatewayDisconnect {
  @WebSocketServer()
  private server: Socket;
  private _rooms: Map<
    string,
    { owner: string; pieces: Piece[]; status: string }
  > = new Map();
  private _clients: Map<string, Player> = new Map();

  handleConnection(@ConnectedSocket() socket: Socket): void {
    console.log("Client connected with id: ", socket.id);
    console.debug(this._clients.size);
  }

  handleDisconnect(@ConnectedSocket() socket: Socket): void {
    this._clients.delete(socket.id);
    console.log("Client disconnected with id: ", socket.id);
    this._rooms.forEach((room, key) => {
      let nbPlayers = 0;
      this._clients.forEach((player) => {
        if (player.room === key) {
          nbPlayers++;
        }
      });
      if (nbPlayers === 0) {
        this._rooms.delete(key);
      }
    });
  }

  @SubscribeMessage("getOwnerByRoomName")
  handleGetRoomByOwner(
    @ConnectedSocket() socket: Socket,
    @MessageBody() body: any,
  ) {
    const owner = this._rooms.get(body.room).owner;
    socket.emit("owner", { owner: owner });
  }

  @SubscribeMessage("join")
  handleJoinGame(
    @ConnectedSocket() socket: Socket,
    @MessageBody() body: any,
  ): void {
    let isValid = true;
    if (!this._rooms.has(body.room)) {
      this._rooms.set(body.room, {
        owner: body.username,
        pieces: PieceFactory.createBagPieces(),
        status: "waiting",
      });
    }
    const players = Array.from(this._clients.values()).filter(
      (player) => player.room === body.room,
    );
    for (const player of players) {
      if (player.name === body.username) {
        isValid = false;
        console.log(
          `Client ${body.username} can't join a user with same username is already in game`,
        );
      }
    }
    if (isValid)
      this._clients.set(
        socket.id,
        new Player(body.username, socket, body.room),
      );
    console.log(`Client ${body.username} joined game with id: ${body.room}`);
  }

  private sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

  private increaseUnavailableLines = (
    players: Player[],
    nbLinesCleared: number,
  ) => {
    while (nbLinesCleared > 0) {
      console.log(
        "players",
        players.map((p) => p.name),
      );
      players.forEach((player: Player) => {
        try {
          console.log("player " + player.name + " unavailableLines++");
          player.game.increaseUnavailableLines();
        } catch (error) {
          console.error(error);
        }
      });
      nbLinesCleared--;
    }
  };

  emitPreviewBoard = (player: Player) => {
    const previewString = player.game.previewBoard.map((row) =>
      row.map((cell) => {
        console.log(cell);
        let colorString = cell.toString(16);
        while (colorString.length < 8) {
          colorString = "0" + colorString;
        }
        return colorString;
      }),
    );
    player.socket.emit("previewBoard", {
      board: previewString,
    });
  };

  @SubscribeMessage("startGame")
  async handleStartGame(@ConnectedSocket() socket: Socket): Promise<void> {
    const currentPlayer = this._clients.get(socket.id);
    if (!currentPlayer) {
      throw new Error("Player not found");
    }
    if (!this._rooms.has(currentPlayer.room)) {
      throw new Error("Room not found");
    }

    if (this._rooms.get(currentPlayer.room).owner !== currentPlayer.name) {
      throw new Error("You are not the owner of the room");
    }

    if (this._rooms.get(currentPlayer.room).status === "playing") {
      throw new Error("Game already started, please wait");
    }

    this._rooms.get(currentPlayer.room).status = "playing";
    let players = Array.from(this._clients.values()).filter(
      (player) => player.room === currentPlayer.room,
    );
    players.forEach((player) => {
      player.startGame(
        new Game(
          this._rooms.get(currentPlayer.room).pieces[0],
          this._rooms.get(currentPlayer.room).pieces[1],
        ),
      );
      player.socket.emit("nextPiece", {
        nextPiece: player.game.nextPiece.nextPiecePreview,
      });
    });
    while (true) {
      const mapBoard = [];
      if (players.length === 0) {
        this._rooms.get(currentPlayer.room).status = "waiting";
        return;
      }
      for (const player of players) {
        const previewString = player.game.spectra.map((row) =>
          row.map((cell) => {
            let colorString = cell.toString(16);
            while (colorString.length < 8) {
              colorString = "0" + colorString;
            }
            return colorString;
          }),
        );
        if (
          player.game.nbPiecePlaced >=
          this._rooms.get(currentPlayer.room).pieces.length / 2
        ) {
          this._rooms.get(currentPlayer.room).pieces = this._rooms
            .get(currentPlayer.room)
            .pieces.concat(PieceFactory.createBagPieces());
        }
        mapBoard.push({ name: player.name, map: previewString });
      }
      let playerToRemove: Player = null;
      for (const player of players) {
        if (player.game.nextPiece === null) {
          player.game.nextPiece = this._rooms.get(currentPlayer.room).pieces[
            player.game.nbPiecePlaced + 2
          ];
          player.game.nextPieceArray;
          player.socket.emit("nextPiece", {
            nextPiece: player.game.nextPiece.nextPiecePreview,
          });
        }
        if (player.game.isGameOver() && players.length === 1) {
          player.socket.emit("winner");
          playerToRemove = player;
        } else if (player.game.isGameOver()) {
          player.socket.emit("gameOver");
          playerToRemove = player;
        }
        player.socket.emit("spectraBoard", mapBoard);
      }
      if (playerToRemove) {
        playerToRemove.endGame();
        players = players.filter((p) => p.name !== playerToRemove.name);
      }

      if (players.length === 0) {
        return;
      }

      players.forEach((player) => {
        this.emitPreviewBoard(player);
      });

      players.forEach((player) => {
        const nbLinesCleared = player.game.moveDown();
        if (nbLinesCleared > 0) {
          console.log("player " + player.name + " send unavailableLines");
        }
        this.increaseUnavailableLines(
          players.filter((p) => p.name !== player.name),
          nbLinesCleared,
        );
      });
      await this.sleep(500);
    }
  }

  @SubscribeMessage("rotate")
  handleRotate(@ConnectedSocket() socket: Socket): void {
    const player = this._clients.get(socket.id);

    if (!player) {
      throw new Error("Player not found");
    }

    if (!player.game) {
      throw new Error("Player not in game");
    }
    player.game.rotate();
    this.emitPreviewBoard(player);
  }

  @SubscribeMessage("moveLeft")
  handleMoveLeft(@ConnectedSocket() socket: Socket): void {
    const player = this._clients.get(socket.id);

    if (!player) {
      throw new Error("Player not found");
    }

    if (!player.game) {
      throw new Error("Player not in game");
    }
    player.game.moveLeft();
    this.emitPreviewBoard(player);
  }

  @SubscribeMessage("moveRight")
  handleMoveRight(@ConnectedSocket() socket: Socket): void {
    const player = this._clients.get(socket.id);

    if (!player) {
      throw new Error("Player not found");
    }

    if (!player.game) {
      throw new Error("Player not in game");
    }
    player.game.moveRight();
    this.emitPreviewBoard(player);
  }

  @SubscribeMessage("moveDown")
  handleMoveDown(@ConnectedSocket() socket: Socket): void {
    const player = this._clients.get(socket.id);

    if (!player) {
      throw new Error("Player not found");
    }

    if (!player.game) {
      throw new Error("Player not in game");
    }
    const nbLinesCleared = player.game.moveDown();
    if (nbLinesCleared > 0) {
      console.log("player " + player.name + " send unavailableLines");
    }
    this.increaseUnavailableLines(
      Array.from(this._clients.values()).filter(
        (p) => p.room === player.room && p.name !== player.name,
      ),
      nbLinesCleared,
    );
    this.emitPreviewBoard(player);
  }

  @SubscribeMessage("drop")
  handleDrop(@ConnectedSocket() socket: Socket): void {
    const player = this._clients.get(socket.id);

    if (!player) {
      throw new Error("Player not found");
    }

    if (!player.game) {
      throw new Error("Player not in game");
    }
    const nbLinesCleared = player.game.drop();
    if (nbLinesCleared > 0) {
      console.log("player " + player.name + " send unavailableLines");
    }
    this.increaseUnavailableLines(
      Array.from(this._clients.values()).filter(
        (p) => p.room === player.room && p.name !== player.name,
      ),
      nbLinesCleared,
    );
    this.emitPreviewBoard(player);
  }
}
