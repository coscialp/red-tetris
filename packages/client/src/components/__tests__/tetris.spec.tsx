import { act, fireEvent, render, screen } from "@testing-library/react";
import Tetris from "../tetris"; // Update the path accordingly
import { Provider } from "react-redux";
// @ts-ignore
import configureStore from "redux-mock-store";
import { SocketActionTypes } from "../../middlewares/socket";

const mockStore = configureStore([]);

describe("Tetris Component", () => {
  it("renders waiting message when grid is empty", () => {
    const store = mockStore({
      rootReducer: {
        me: "player1",
        owner: "player2",
        isPlaying: false,
      },
    });

    render(
      <Provider store={store}>
        <Tetris me="player1" owner="player2" isPlaying={false} />
      </Provider>,
    );

    expect(screen.getByText(/Waiting for the owner to start the game.../i)).toBeDefined();
  });

  it("renders start message when grid is empty and user is the owner", () => {
    const store = mockStore({
      rootReducer: {
        me: "player1",
        owner: "player1",
        isPlaying: false,
      },
    });

    render(
      <Provider store={store}>
        <Tetris me="player1" owner="player1" isPlaying={false} />
      </Provider>,
    );

    expect(screen.getByText(/Press Start to launch the game/i)).toBeDefined();
  });

  it("renders board when grid is not empty", async () => {
    const store = mockStore({
      rootReducer: {
        me: "player1",
        owner: "player1",
        isPlaying: true,
      },
    });

    const mockGrid = [
      ["00000000", "FF0000FF", "00FF00FF"],
      ["00000000", "00FF00FF", "FF0000FF"],
    ];

    //@ts-ignore
    let tetris = null;
    act(() => {
      tetris = render(
        <Provider store={store}>
          <Tetris me="player1" owner="player1" isPlaying={true} />
        </Provider>,
      );
    });

    // Mock the 'previewBoard' event
    fireEvent(
      window,
      new MessageEvent("previewBoard", {
        data: JSON.stringify({
          event: SocketActionTypes.ON,
          data: { board: mockGrid },
        }),
      }),
    );

    tetris!.board = mockGrid;

    // Use 'container' to check the rendered structure
    expect(tetris!.container).toMatchSnapshot();
  });
});