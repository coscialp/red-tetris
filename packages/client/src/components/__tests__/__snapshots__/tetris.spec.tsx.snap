// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tetris Component renders board when grid is not empty 1`] = `
<div>
  <div
    class="board"
  >
    <div
      class="row"
    >
      <div
        class="cell empty"
      />
      <div
        class="cell full"
        style="background-color: rgb(230, 0, 0); border-color: #ff0000ff;"
      />
      <div
        class="cell full"
        style="background-color: rgb(0, 234, 0); border-color: #00ff00ff;"
      />
    </div>
    <div
      class="row"
    >
      <div
        class="cell empty"
      />
      <div
        class="cell full"
        style="background-color: rgb(0, 234, 0); border-color: #00ff00ff;"
      />
      <div
        class="cell full"
        style="background-color: rgb(230, 0, 0); border-color: #ff0000ff;"
      />
    </div>
  </div>
</div>
`;
